Reimplement the DecorationManager

pierremoreau opened the pull request:
This reimplementation fixes several issues when removing decorations associated to an ID (partially addresses #1174 and gives tools for fixing #898), as well as making it easier to remove groups; a few additional tests have been added.

`DecorationManager::RemoveDecoration()` will still not delete dead decorations it created, but I do not think it is its job either; given the following input

```
OpCapability Shader
OpCapability Linkage
OpMemoryModel Logical GLSL450
OpDecorate %2 Restrict
%2      = OpDecorationGroup
OpGroupDecorate %2 %1 %3
OpDecorate %4 Invariant
%4      = OpDecorationGroup
OpGroupDecorate %4 %2
%uint   = OpTypeInt 32 0
%1      = OpVariable %uint Uniform
%3      = OpVariable %uint Uniform
```

which of the following two outputs would you expect `RemoveDecoration(2)` to produce:

```
OpCapability Shader
OpCapability Linkage
OpMemoryModel Logical GLSL450
%uint = OpTypeInt 32 0
%1 = OpVariable %uint Uniform
%3 = OpVariable %uint Uniform
```

or

```
OpCapability Shader
OpCapability Linkage
OpMemoryModel Logical GLSL450
OpDecorate %4 Invariant
%4      = OpDecorationGroup
%uint   = OpTypeInt 32 0
%1      = OpVariable %uint Uniform
%3      = OpVariable %uint Uniform
```

alan-baker commented:
Your example is invalid because OpGroupDecorate cannot target an OpDecorationGroup. Does that simplify the logic required in the DecorationManager?

pierremoreau commented:
That definitely would make some things simpler, however there is nothing in the spec preventing OpGroupDecorate from targeting other groups. @dneto0 brought up this point in #924 and was going to bring it up to the SPIR-V working group, but I haven’t heard back since.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -18,14 +18,91 @@
+        if (i < last_operand_index)
+          inst->GetInOperand(i + 1u) =
+              inst->GetInOperand(last_operand_index + 1u);
+        inst->RemoveInOperand(last_operand_index + 1u);
```
The use if `pred` seem inconsistent.  If the decoration is applied through a group you do not check it.  That does not seems right.

pierremoreau replied:
You are right, I completely missed that case.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -18,14 +18,91 @@
+  // Schedule all instructions applying the group for removal if instructed as
+  // such by |pred|.
+  for (ir::Instruction* inst : decorations_info.decorate_insts)
+    insts_to_kill.push_back(inst);
```
Can you do this if `pred` returned false for any of the direct decorations?

pierremoreau replied:
No, you are right, I should check that it has no direct decorations left.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -18,14 +18,91 @@
+  }
+
+  // Remove the OpDecorationGroup defining this group.
+  if (is_group) context->KillInst(context->get_def_use_mgr()->GetDef(id));
```
You cannot delete this unless all uses of `id` are also killed, but what if `pred` returned false for something?

pierremoreau replied:
This should be in the if statement right above it, to guard against that eventuality.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -180,46 +235,42 @@ template <typename T>
-      return std::vector<T>();
-    }
+  // Process the decorations of all groups applied to |id|.
+  std::stack<uint32_t> groups_to_process;
```
Just use an std::vector.  We want to avoid more uses of std::stack, std::queue, std::list, and std::deueque.  They perform very poorly on windows.

pierremoreau replied:
Okay, I’ll do the change to std::vector.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -180,46 +235,42 @@ template <typename T>
+  for (const ir::Instruction* inst : target_data.indirect_decorations)
+    groups_to_process.push(inst->GetSingleWordInOperand(0u));
+
+  std::unordered_set<uint32_t> group_ids;
```
Do you think there will be duplicates?  If not std::vector would again be the better data structure.  You don't seem to do any searches.

pierremoreau replied:
I think there could be duplicates, but I’ll try to come up with such a SPIR-V example to check whether it is possible or not.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -260,16 +311,25 @@ void DecorationManager::ForEachDecoration(
+    // simply clone decoration and change |target-id| to |to|
+    std::unique_ptr<ir::Instruction> new_inst(inst->Clone(module_->context()));
+    new_inst->SetInOperand(0, {to});
+    id_to_decoration_insts_[to].direct_decorations.push_back(new_inst.get());
```
Doing the same hash lookup for every iteration.  Might want to move that out of the loop.

pierremoreau replied:
I’ll store a reference to |to|’s `TargetData` outside of the loop, and use it in this loop and in others further down in the function.

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -260,16 +311,25 @@ void DecorationManager::ForEachDecoration(
 void DecorationManager::CloneDecorations(
     uint32_t from, uint32_t to, std::function<void(ir::Instruction&, bool)> f) {
```
Given that the decoration now knows about the context, we might be able to remove the call back `f`.  Just have the context analyze the new instructions.

pierremoreau replied:
I’ll do that.

s-perron commented on source/opt/decoration_manager.h:
```diff
@@ -35,10 +36,16 @@ class DecorationManager {
-  void RemoveDecorationsFrom(uint32_t id);
+  // Removes all decorations from |id| for which |pred| returns true.
+  // |id| should not be a group ID.
+  void RemoveDecorationsFrom(uint32_t id,
```
Can you explain why you are adding `pred`?  It will make the implementation much more difficult, so we should have a reason.

pierremoreau replied:
The reason for adding pred was explained in #898, more specifically in @dneto0’s comment https://github.com/KhronosGroup/SPIRV-Tools/issues/898#issuecomment-338413513. Currently the linker has some code (which is wrong) that loops over all decorations to remove only LinkageAttributes Import ones, or FuncParamAttr ones. The idea is to have the DecorationManager do it, in a generic way, instead.

pierremoreau commented:
@alan-baker Ah, I missed the update to OpGroupDecorate in SPIR-V 1.2 rev 3 that explicitly says it can’t target groups. I’ll update the code accordingly.

pierremoreau commented:
I forgot to add in one of the commits:

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/924

pierremoreau commented:
Since groups of groups aren’t valid, I could change the code to remove a dead group when removing its last target in RemoveDecoration(). Or should this be left to ADCE, which already handles dead decorations IIRC?

alan-baker commented:
You don't need to remove target-less group decorations. Feel free to leave them for ADCE.

pierremoreau commented:
> You don't need to remove target-less group decorations. Feel free to leave them for ADCE.

Okay, nice. In that case, I think this should also include:

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/1174

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -46,6 +46,21 @@ void DecorationManager::RemoveDecorationsFrom(
+    const uint32_t group_id = inst->GetSingleWordInOperand(0u);
+    const auto group_iter = id_to_decoration_insts_.find(group_id);
+    for (ir::Instruction* group_decoration :
+         group_iter->second.direct_decorations)
```
Please add the braces.  It is easier to see what is in the scope of the for loop.

pierremoreau replied:
Will do

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -46,6 +46,21 @@ void DecorationManager::RemoveDecorationsFrom(
+    // If all decorations should be kept, move to the next group
+    if (group_decorations_to_keep.size() ==
+        group_iter->second.direct_decorations.size())
+      continue;
```
If you are keeping all of the decorations, should you clear `group_decorations_to_keep` so you don't add them below?

pierremoreau replied:
You are correct, I need to do that. I’ll add a test that would fail with the current code, to detect possible regressions and check that the code is properly fixed.

pierremoreau replied:
Actually, looking at the code again, they won’t be added later on (I assume you are thinking of the code [right here](https://github.com/KhronosGroup/SPIRV-Tools/pull/1373/files/91f73ab072b9176e4e539a343d868f005847e8a4#diff-b567013ae7fa0970645039363ebc5cd7R99)): `group_decorations_to_keep` is local to the loop, so when continuing and going to the next loop iteration, `group_decorations_to_keep` will be a new, empty, vector.

I added two new tests to check that behaviour, which uncovered a bug where the medata was not updated with the clone instructions due to calling `AnalyzeDefUse()` (which does not update decorations) instead of `AnalyzeUses()`.
I alos added a test for the `CloneDecorations()` which was suffering from the same issue, as well as from an other bug (the list of direct decorations from groups was modified as we were iterating over it).

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -242,35 +276,26 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
-        inst->GetSingleWordInOperand(1u) == SpvDecorationLinkageAttributes;
-    if (include_linkage || !is_linkage) decorations.push_back(inst);
-  }
+  process_direct_decorations(ids_iter->second.direct_decorations);
```
If there are only direct calls to this function, why use a lambda?

pierremoreau replied:
The lambda is called twice: once on line 291, and once on line 298 inside the loop over groups’ direct decorations.

s-perron replied:
I know it is called twice.  I'm not suggesting it should be inlined.  I'm just wondering why a regular function, where `include_linkage` and `decorations` are passed as parameters,  would not work.

pierremoreau replied:
Ah sorry, I misunderstood your comment. A regular function could work as well, but I am not sure who else would call it, and since there was only one place needing it, a lambda seemed to be the best. Also, I was thinking I could replace `include_linkage` by a predicate function instead.

pierremoreau commented:
Please have another look.
(I added the fixes messages to the last commit.)

@s-perron I addressed your remaining comments, apart from the regular function one. I can make the lambda into a regular function if you want, but I would rather get rid of `include_linkage` and replace it by a predicate, similar to `RemoveDecorationsFrom()`.

s-perron commented:
The lambda is not a big deal.  I we can put it through anyway.  However, even if you change `include_linkage` to an `std::function`, it will not force `process_direct_decorations` to be a lambda.  You can simply pass the predicate function as a parameter to the function.

pierremoreau commented:
>  However, even if you change include_linkage to an std::function, it will not force process_direct_decorations to be a lambda. You can simply pass the predicate function as a parameter to the function.

No, in that case I would even drop the lambda and inline the code, though a regular function would work fine as well, as it would even be used by `RemoveDecorationsFrom()`.

s-perron commented:
Rebased, squashed, and merged as 5bd55f10.

pierremoreau commented:
Thank you all for the comments and reviews. :-)
